Support creation of records while not being assigned to any institution

palagdan opened the pull request:
Resolve partially kbss-cvut/record-manager-ui#302.

palagdan committed:
[kbss-cvut/record-manager-ui#302] Add variable for allowing creation records without being assigned to any institution

palagdan committed:
[kbss-cvut/record-manager-ui#302] Showing also records without institution

palagdan committed:
[kbss-cvut/record-manager-ui#302] Move permission's validation to the separate function and implement record collision validation

palagdan committed:
[kbss-cvut/record-manager-ui#302] Fix and implement new tests

palagdan committed:
[kbss-cvut/record-manager-ui#302] Change records institution after changing user institution

palagdan committed:
[kbss-cvut/record-manager-ui#302] Fix modifiers

blcham commented on src/main/java/cz/cvut/kbss/study/persistence/dao/PatientRecordDao.java:
```diff
@@ -172,11 +172,16 @@ public int getNumberOfProcessedRecords() {
         }
+
+        if(entity.getInstitution() == null){
+            em.clear();
```
?

blcham commented on src/main/java/cz/cvut/kbss/study/persistence/dao/PatientRecordDao.java:
```diff
@@ -304,13 +309,16 @@ private void setQueryParameters(Query query, Map<String, Object> queryParams) {
+                "OPTIONAL { " +
+                "?r ?hasInstitution ?institution . " +
+                "?institution ?hasKey ?institutionKey . " +
+                "} " +
```
This is kind of crazy to read ... how about using multiline strings in JAVA with correct indentation here?

blcham replied:
Refactor it, please, also elsewhere in this file as well ...

palagdan replied:
done

blcham commented on src/main/java/cz/cvut/kbss/study/util/Configuration.java:
```diff
@@ -422,12 +422,25 @@ public static class Records {
         boolean allowedRejectReason;
 
+        /**
+         *  it indicates functionality allowing users to create records without being assigned to any institution.
```
```suggestion
         * Allow users who are not members of any institution to create records.
```

blcham replied:
As i was not sure if the requirement is on the user or the record.

palagdan replied:
done

blcham commented on src/test/java/cz/cvut/kbss/study/rest/PatientRecordControllerTest.java:
```diff
@@ -224,6 +229,22 @@ public void createRecordWithoutInstitutionReturnsResponseStatusBadRequest() thro
     }
 
+    @Test
+    public void createRecordWithoutInstitutionIfItIsAllowedReturnsResponseStatusBadRequest() throws Exception {
```
```suggestion
    public void createRecordWithoutInstitutionIfItIsAllowedReturnsResponseStatusCreated() throws Exception {
```

palagdan replied:
done

blcham committed:
[kbss-cvut/record-manager-ui#302] Improve validation error

blcham commented on src/main/java/cz/cvut/kbss/study/service/repository/RepositoryUserService.java:
```diff
@@ -243,6 +235,71 @@ protected void preUpdate(User instance) {
         }
+    }
+
+    protected void validateRecordsAgainstCollisions(User toUpdate, User original) {
```
@palagdan, look at my last commit. I had the following issues:
- 1) to me, variable names were not correct, e.g,. newInsitutionRecords ==> existingInstitutionRecords
- 2) although not that important -- algorithm was of higher complexity than needed because of comparing every label with every other label, i.e., on average O(N^2) instead of O(N) on average
- 3) IMPORTANT: I wanted to have a more detailed message on errors (although we don't have parametrized so we cannot navigate from the message to actual conflicting records but I would like to get there someday :)
- 4) IMPORTANT: patient records --> records (we have technical debt there :( )

blcham replied:
- I think there might be problem within UI how to show this message due to `\n` in the message and how it is interpreted in HTML (did not test it)

blcham replied:
if you don't like the solution, revert it and address please at least IMPORTANT bullets.

palagdan replied:
I like this solution :)

blcham reviewed (changes requested):
I refactored your code; see the last commit. Have a look and if you don't like it revert it, but i would like to address issues described there.

palagdan committed:
[kbss-cvut/record-manager-ui#302] Indent where clauses

palagdan committed:
[kbss-cvut/record-manager-ui#302] Change environment variable description

palagdan committed:
[kbss-cvut/record-manager-ui#302] Fix test name

palagdan committed:
[kbss-cvut/record-manager-ui#302] Fix Record naming after rebasing

blcham committed:
[kbss-cvut/record-manager-ui#302] Fix record persist/update without institution

Records without an institution will throw NPE during save/update instead of being allowed through.

blcham committed:
[kbss-cvut/record-manager-ui#302] Fix pagination results

Records without an institution were excluded from paginated results because
?hasInstitution ?institution was a required triple. Made it OPTIONAL and
wrapped the GRAPH ?institutionGraph lookup in OPTIONAL as well.

blcham committed:
[kbss-cvut/record-manager-ui#302] Fix institution comparison in preUpdate to use URI equality

Reference equality (!=) always evaluated to true when comparing deserialized
request body institution against a DB-fetched institution, causing records to
be unnecessarily migrated on every user update. Compare by URI instead.

blcham committed:
[kbss-cvut/record-manager-ui#302] Rename changePatientRecordsInstitution to changeRecordsInstitution

blcham committed:
[kbss-cvut/record-manager-ui#302] Allow user to view own record without having read permissions

blcham reviewed (changes requested):
Reviewed and it looks ok, but testing it locally showed multiple issues that needs to be solved first, follow the test scenario:
- 1. substitute RM with current branch
- 2. add user without institution
- 3. add a record (should be success) (related issue -- https://github.com/kbss-cvut/record-manager/issues/106)
- 4. set `records.allowedCreationWithoutInstitution: false`
- 5. add a record (should be fail)
- 6. set user an organization (his records should migrate as well)
